File: Android/Udpsteering/app/src/main/java/com/beamng/remotecontrol/input/ButtonInputHandler.java
package com.beamng.remotecontrol.input;

import android.content.Context;
import android.view.MotionEvent;
import android.view.View;

import com.beamng.remotecontrol.settings.SettingsManager;

/**
 * Buton tabanlı direksiyon kontrolü.
 * Sol/Sağ butonlarına basarak araba yönlendirilir.
 */
public class ButtonInputHandler implements SteeringInputHandler {

    private final SettingsManager settings;
    
    private float currentSteering = 0f;
    private boolean leftPressed = false;
    private boolean rightPressed = false;
    
    public ButtonInputHandler(Context context) {
        this.settings = SettingsManager.getInstance(context);
    }
    
    @Override
    public float getSteeringValue() {
        float sensitivity = settings.getSensitivity();
        return currentSteering * sensitivity;
    }
    
    @Override
    public void start() {
        // Butonlar için özel başlatma gerekmez
    }
    
    @Override
    public void stop() {
        currentSteering = 0f;
        leftPressed = false;
        rightPressed = false;
    }
    
    @Override
    public boolean requiresUIControls() {
        return true; // Butonlar gerekiyor
    }
    
    /**
     * Sol butona bağlanacak OnTouchListener.
     */
    public View.OnTouchListener getLeftButtonListener() {
        return (v, event) -> {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    leftPressed = true;
                    updateSteering();
                    break;
                case MotionEvent.ACTION_UP:
                case MotionEvent.ACTION_CANCEL:
                    leftPressed = false;
                    updateSteering();
                    break;
            }
            return true;
        };
    }
    
    /**
     * Sağ butona bağlanacak OnTouchListener.
     */
    public View.OnTouchListener getRightButtonListener() {
        return (v, event) -> {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    rightPressed = true;
                    updateSteering();
                    break;
                case MotionEvent.ACTION_UP:
                case MotionEvent.ACTION_CANCEL:
                    rightPressed = false;
                    updateSteering();
                    break;
            }
            return true;
        };
    }
    
    private void updateSteering() {
        if (leftPressed && !rightPressed) {
            currentSteering = -1f; // Tam sola
        } else if (rightPressed && !leftPressed) {
            currentSteering = 1f; // Tam sağa
        } else {
            currentSteering = 0f; // Ortada veya ikisi de basılı
        }
    }
}

File: Android/Udpsteering/app/src/main/java/com/beamng/remotecontrol/input/ButtonInputHandler.kt
package com.beamng.remotecontrol.input

import android.content.Context
import android.os.Handler
import android.os.Looper

import com.beamng.remotecontrol.settings.SettingsManager

/**
 * Button-based steering control with ramp-up.
 * Holding a button gradually increases steering from 0 to 1 over ~350ms,
 * giving analog-like feel from digital buttons.
 */
class ButtonInputHandler(context: Context) : SteeringInputHandler {

    private val settings: SettingsManager = SettingsManager.getInstance(context)
    private val handler = Handler(Looper.getMainLooper())

    @Volatile
    private var currentSteering = 0f
    private var leftPressed = false
    private var rightPressed = false
    private var targetSteering = 0f

    private val rampRunnable = object : Runnable {
        override fun run() {
            if (!leftPressed && !rightPressed) {
                currentSteering = 0f
                return
            }

            // Ramp toward target
            if (currentSteering < targetSteering) {
                currentSteering = minOf(targetSteering, currentSteering + RAMP_STEP)
            } else if (currentSteering > targetSteering) {
                currentSteering = maxOf(targetSteering, currentSteering - RAMP_STEP)
            }

            // Keep ticking while buttons are held
            if (leftPressed || rightPressed) {
                handler.postDelayed(this, RAMP_TICK_MS)
            }
        }
    }

    override fun getSteeringValue(): Float = currentSteering

    override fun start() {
    }

    override fun stop() {
        handler.removeCallbacks(rampRunnable)
        currentSteering = 0f
        targetSteering = 0f
        leftPressed = false
        rightPressed = false
    }

    override fun requiresUIControls(): Boolean = true

    fun pressLeft(pressed: Boolean) {
        leftPressed = pressed
        updateTarget()
        if (pressed) startRamp()
    }

    fun pressRight(pressed: Boolean) {
        rightPressed = pressed
        updateTarget()
        if (pressed) startRamp()
    }

    private fun updateTarget() {
        if (leftPressed && !rightPressed) {
            targetSteering = -1f
        } else if (rightPressed && !leftPressed) {
            targetSteering = 1f
        } else {
            targetSteering = 0f
            currentSteering = 0f // Instant center on release
        }
    }

    private fun startRamp() {
        handler.removeCallbacks(rampRunnable)
        handler.post(rampRunnable)
    }

    companion object {
        private const val RAMP_TICK_MS = 50L
        private const val RAMP_STEP = 0.15f // ~7 ticks (350ms) to reach 1.0
    }
}
